File: done/232_queue.py
##!/usr/bin/env python
'''
    @author [mst]
    @file    232_queue.py
    @brief  leetcode problems series
    232. Implement Queue using Stacks
    Implement a first in first out (FIFO) queue using only two stacks.
    The implemented queue should support all the functions of a normal queue (push, peek, pop, and empty).

    features, changelog:
    -2022.12: -submission

    @version 0.1 2022.12
'''

# I think I've cheated here by exploiting py3 lists :|
#
# sub: 99%T 76%S
class MyQueue:

    fifo = []

    def __init__(self):
        self.fifo = []

    def push(self, x: int) -> None:
        self.fifo.append(x)

    def pop(self) -> int:
        if len(self.fifo) > 0:
            x = self.fifo.pop(0)
            return x
        return -1

    def peek(self) -> int:
        if len(self.fifo) > 0:
            return self.fifo[0]
        return -1

    def empty(self) -> bool:
        return len(self.fifo) == 0



################## DRIVER
def main():
    print ("[mst]")

    obj = MyQueue()
    obj.push(1)
    param_2 = obj.pop()
    param_3 = obj.peek()
    param_4 = obj.empty()



if __name__ == ("__main__"):
    main()

File: done/232_queue.cpp
/**
    @author [mst]
    @brief  leetcode series
    232. Implement Queue using Stacks
    Implement a first in first out (FIFO) queue using only two stacks.
    The implemented queue should support all the functions of a normal queue (push, peek, pop, and empty).

    gains:
    -stack and queue basics


    @version 0.1 2023.06
*/



////////////////// LIBS
#include <iostream>     // usage of console prints
#include <vector>
#include <algorithm>    // usage of sort
#include <cstring>      // use of memset
#include <numeric>      // use of accumulate
using namespace std;


////////////////// DECL_IMPL

//
//
// sub 100 15
class MyQueue {
private:
  const int MAX_SIZE = 100;
  vector<int> stackIn;
  vector<int> stackOut;
public:
    MyQueue() {
      stackIn.reserve( MAX_SIZE );
      stackOut.reserve( MAX_SIZE );
    }

    void push(int x) {
      stackIn.push_back(x);
    }

    int pop() {
      if (stackOut.empty()) {
        while (!stackIn.empty()) {
          stackOut.push_back(stackIn.back());
          stackIn.pop_back();
        }
      }
      int ans = stackOut.back();
      stackOut.pop_back();
      return ans;
    }

    int peek() {
      if (stackOut.empty()) {
        while (!stackIn.empty()) {
          stackOut.push_back(stackIn.back());
          stackIn.pop_back();
        }
      }
      int ans = stackOut.back();
      return ans;


    }

    bool empty() {
      return (max(stackIn.size(),stackOut.size()) == 0);
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

////////////////// DRIVER
int main()
{
    int x = 4;
    MyQueue* obj = new MyQueue();
    obj->push(x);
    int param_2 = obj->pop();
    int param_3 = obj->peek();
    bool param_4 = obj->empty();

    return 0;
}